### ISSUE TITLE
Dotted-path to a callable for rate= argument does not work

### ISSUE THREAD

**Issue by shubham17sm** [jsocol/django-ratelimit#204]
Environment:

Request Method: GET
Request URL: http://127.0.0.1:8000/convert/75/eng/

Django Version: 2.1.5
Python Version: 3.8.2
Installed Applications:
['django.contrib.admin',
 'django.contrib.auth',
 'django.contrib.contenttypes',
 'django.contrib.sessions',
 'django.contrib.messages',
 'django.contrib.staticfiles',
 'django.contrib.sites',
 'myapp',
 'allauth',
 'allauth.account',
 'allauth.socialaccount',
 'debug_toolbar']
Installed Middleware:
['django.middleware.security.SecurityMiddleware',
 'django.contrib.sessions.middleware.SessionMiddleware',
 'django.middleware.common.CommonMiddleware',
 'django.middleware.csrf.CsrfViewMiddleware',
 'django.contrib.auth.middleware.AuthenticationMiddleware',
 'django.contrib.messages.middleware.MessageMiddleware',
 'django.middleware.clickjacking.XFrameOptionsMiddleware',
 'debug_toolbar.middleware.DebugToolbarMiddleware']

Traceback:

File "/home/dev/projects/findr/lib/python3.8/site-packages/django/core/handlers/exception.py" in inner
  34.             response = get_response(request)

File "/home/dev/projects/findr/lib/python3.8/site-packages/django/core/handlers/base.py" in _get_response
  126.                 response = self.process_exception_by_middleware(e, request)

File "/home/dev/projects/findr/lib/python3.8/site-packages/django/core/handlers/base.py" in _get_response
  124.                 response = wrapped_callback(request, *callback_args, **callback_kwargs)

File "/home/dev/projects/findr/lib/python3.8/site-packages/django/contrib/auth/decorators.py" in _wrapped_view
  21.                 return view_func(request, *args, **kwargs)

File "/home/dev/projects/findr/lib/python3.8/site-packages/ratelimit/decorators.py" in _wrapped
  18.             ratelimited = is_ratelimited(request=request, group=group, fn=fn,

File "/home/dev/projects/findr/lib/python3.8/site-packages/ratelimit/core.py" in is_ratelimited
  119.     usage = get_usage(request, group, fn, key, rate, method, increment)

File "/home/dev/projects/findr/lib/python3.8/site-packages/ratelimit/core.py" in get_usage
  162.     limit, period = _split_rate(rate)

File "/home/dev/projects/findr/lib/python3.8/site-packages/ratelimit/core.py" in _split_rate
  81.     count, multi, period = rate_re.match(rate).groups()

Exception Type: AttributeError at /convert/75/eng/
Exception Value: 'NoneType' object has no attribute 'groups'

**Comment by adamchainz** [comment]
The `rate` you're passing doesn't match `ratelimit`'s format. See https://django-ratelimit.readthedocs.io/en/stable/rates.html#rates-chapter

**Comment by jsocol** [comment]
This seems like an opportunity for a better error message. I think there’s a django exception for configuration, right?

**Comment by adamchainz** [comment]
Yes, `ImproperlyConfigured`: https://github.com/django/django/blob/c1f8d87bb0ac72b213bf31e672bd34c93bda6b18/django/core/exceptions.py#L86

**Comment by shubham17sm** [comment]
Still can't figure out 
models.py
class Ratelimit(models.Model):
    group = models.CharField(max_length=100, db_index=True)
    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, null=True)  # One option for "default"
    rate = models.CharField(max_length=100)

    @classmethod
    def get(cls, group, user=None):
        # use cache if possible
        try:
            return cls.objects.get(group=group, user=user) 
        except cls.DoesNotExist:
            return cls.objects.get(group=group, user=None)

views.py
@login_required
@ratelimit(group='', key='user',
           rate='myapp.ratelimits.per_user')
def convert_image(request, pk, lang):
    session_key = get_session_key(request)
    
    try:
        get_image = ImageToConvert.objects.get(pk=pk, session_key=session_key)

        print(get_image.file.url)

        command = 'tesseract ' + get_image.file.path + \
            ' media/documents/output -l ' + lang

        process = subprocess.Popen(command.split(), stdout=subprocess.PIPE)

        with open('media/documents/output.txt', 'r') as uploadedFile:
            converted_data = uploadedFile.read().replace('\n', '<br />')

        get_image.converted_text = converted_data
        get_image.save()

        messages.success(request, "Successfully converted file")
        return redirect('/')
    except Exception as e:
        print(e)
        print('error converting the file')
        messages.warning(request, "Error converting the file")
        return redirect('/')

ratelimits.py
from .models import Ratelimit
def per_user(group, request):
    if request.user.is_authenticated:
        return Ratelimit.get(group, request.user)
    return Ratelimit.get(group, rate)

In admin panel i made an object as:
group: free
user: admin
rate: 10/d

Please let me know if i made some mistake?

**Comment by shubham17sm** [comment]
Still can't figure out: 
models.py 
```
class Ratelimit(models.Model):
    group = models.CharField(max_length=100, db_index=True)
    user = models.ForeignKey(settings.AUTH_USER_MODEL, on_delete=models.CASCADE, null=True)  # One option for "default"
    rate = models.CharField(max_length=100)

    @classmethod
    def get(cls, group, user=None):
        # use cache if possible
        try:
            return cls.objects.get(group=group, user=user) 
        except cls.DoesNotExist:
            return cls.objects.get(group=group, user=None)
```
ratelimits.py 
```
from .models import Ratelimit
def per_user(group, request):
    if request.user.is_authenticated:
        return Ratelimit.get(group, request.user)
    return Ratelimit.get(group, rate)
```

views.py
```@login_required
@ratelimit(group='', key='user',
           rate='myapp.ratelimits.per_user')
def convert_image(request, pk, lang):
    session_key = get_session_key(request)
    
    try:
        get_image = ImageToConvert.objects.get(pk=pk, session_key=session_key)

        print(get_image.file.url)

        command = 'tesseract ' + get_image.file.path + \
            ' media/documents/output -l ' + lang

        process = subprocess.Popen(command.split(), stdout=subprocess.PIPE)

        with open('media/documents/output.txt', 'r') as uploadedFile:
            converted_data = uploadedFile.read().replace('\n', '<br />')

        get_image.converted_text = converted_data
        get_image.save()

        messages.success(request, "Successfully converted file")
        return redirect('/')
    except Exception as e:
        print(e)
        print('error converting the file')
        messages.warning(request, "Error converting the file")
        return redirect('/')
```

please let me know if i made some mistake?

**Comment by jsocol** [comment]
Thank you for that code snippet, this is definitely a bug in ratelimit.

in https://github.com/jsocol/django-ratelimit/blob/main/ratelimit/core.py#L158-L159
```py
if callable(rate):
    rate = rate(group, request)
```
there's no equivalent to https://github.com/jsocol/django-ratelimit/blob/main/ratelimit/core.py#L175-L176 for callable keys
```py
if '.' in key:
    keyfn = import_string(key)
```

**Comment by jsocol** [comment]
Also, in the docs, it says we assume it's a dotted path if there's no `/` in the string, but we should probably update the doc, and the logic, to be "if there's a `'.'` in the string", for consistency with callable keys.

**Comment by shubham17sm** [comment]
Yes document is not very clear about implementation.
so i am assuming you are trying to say there is no solution as of now to solve the error that i was getting?

**Comment by jsocol** [comment]
> Yes document is not very clear about implementation.

I think the callable keys docs are better, we can probably follow that example.

> so i am assuming you are trying to say there is no solution as of now to solve the error that i was getting?

One thing that might work is importing `per_user` and passing it directly, e.g. `@ratelimit(rate=per_user, ...)`, instead of the dotted path. That will work **if** it does not cause a circular dependency or attempt to load a model too early in the Django start-up process.

Otherwise, you are correct. I can probably patch this tonight (GMT-4) and get a release up.

**Comment by shubham17sm** [comment]
@jsocol Thankyou so much for your help. I really appreciate your help. I tried using importing `per_user` in views.py but i didn't work its giving `expected string or bytes-like object`. I will wait until to patch and release this changes, Thankyou.

**Comment by jsocol** [comment]
@shubham17sm I think `expected string or bytes-like object` is because your `per_user` function is returning `Ratelimit` objects, not strings. Once you have the `obj = Ratelimit.get(group, user)` object, you need to do something like

```py
obj = Ratelimit.get(group, user)
if obj is None:
    return None
return obj.rate
```

I realize this is wrong in the cookbook section of the docs, I'll fix that, too!

**Comment by jsocol** [comment]
(It can also return tuples, e.g. `(10, 86400)`, but that model assumes rates are stored as strings e.g. `10/d`.)

**Comment by shubham17sm** [comment]
It worked!!
```
from .models import Ratelimit

def per_user(group, request):
    obj = Ratelimit.get(group, user=request.user)
    if obj is None:
        return None
    return obj.rate  
```

**Comment by adamchainz** [comment]
IMO `rate` taking a dotted path could be deprecated. It's useful in Django settings for hooking up things pre-import, but generally one can structure stuff like this without it. The rate callable can always use inner imports to avoid circular dependencies.

**Comment by shubham17sm** [comment]
Also look into, as i set my rate=1/d,  i.e views should only work for one time in a day. but as i upload and try to convert image twice it still rendering the view, it should have given 403 error but its not giving

**Comment by shubham17sm** [comment]
@ratelimit(group='free', key='user', rate=per_user, block=True)
needs to change in per-user ratelimit in cookbook

**Comment by shubham17sm** [comment]
can we do this to get if multiple groups are present?
```
def get_group(group, request):
    obj = Ratelimit.get(group, user=request.user)
    if obj is None:
        return None
    return obj.group
```

**Comment by jsocol** [comment]
> IMO `rate` taking a dotted path could be deprecated. It's useful in Django settings for hooking up things pre-import, but generally one can structure stuff like this without it. The rate callable can always use inner imports to avoid circular dependencies.

That's fair, but I don't know if deprecating it is high priority? Removing it is a major version update and I'd rather fix the bug in the documented behavior without a major bump right now. Maybe add a deprecation warning to it in the future.

**Comment by jsocol** [comment]
> Also look into, as i set my rate=1/d, i.e views should only work for one time in a day. but as i upload and try to convert image twice it still rendering the view, it should have given 403 error but its not giving

Can you file a separate issue for this? And please include whether it is returning a 403 on the 3rd attempt, or never?

**Comment by jsocol** [comment]
> can we do this to get if multiple groups are present?
> 
> ```
> def get_group(group, request):
>     obj = Ratelimit.get(group, user=request.user)
>     if obj is None:
>         return None
>     return obj.group
> ```

I'm not sure what you mean "multiple groups". Each ratelimit can only belong to a single group (which is not the same as a user group, which is a many-to-many relationship). Ratelimit groups allow you to link multiple endpoints into a single limit.

**Comment by adamchainz** [comment]
> That's fair, but I don't know if deprecating it is high priority?

Absolutely not, just wanted to float the opinion :) I'd do it if I had time.

**Comment by jsocol** [comment]
This is fixed in v3.0.1 which is on pypi now!
